docs/rfcs/025-generation-numbers.md

```diff
@@ -0,0 +1,468 @@
+coupling:
+
+- Control plane must not "steal" a tenant from a node (by attaching to another node without detaching
+  from the first node) without also incrementing the original node's generation number, which
+  effectively marks the original node as dead.
+- This enables nodes to trust that if their generation ID is still current, then all their
+  attachments are as well.
+
+### Object Key Changes
+
+All object keys (layer objects and index objects) will have a suffix as follows:
+
+```
+   <node id>-<attachment generation>-<node generation>
+```
+
+To avoid unreasonably long keys, we may make some reasonable assumptions about max lengths
+of these integers:
+
+- 32 bit attachment generation (this easily accommodates reasonable worst case scenarios like
+  a node restarting every minute for 5 years)
+- 16 bit node ID (this doesn't have to accommodate all pageservers ever, just all the pageservers
+  that act together in a logical cluster)
+- 32 bit node generation (as with attachment generation, this can handle even very frequent restarts)
+
+So an example suffix for generation 1 of node 0 might look like:
+
+```
+   00000001-0000-00000001
+```
+
+This suffix provides a lot of flexibility in how we may handle migration/restarts/failover
+at runtime:
+
+- We can replace a pageserver with another physical node having the same pageserver ID, and
+  the pageserver ID in the name will protect from conflicting writes.
+- Two pageservers may have the same tenant attached without risking stepping on each others
+  writes.
+- Ultimately, pageservers might self-coordinate such that a single "attachment" refers to
+  multiple pageservers, and they may internally do HA without troubling the control
+  plane to do new attachments, as the node generation makes this safe.
+
+### Deletion Changes
+
+While writes are de-conflicted by writers always using their own generation number in the key,
+deletions are slightly more challenging: if a pageserver A is isolated, and the true active node is
+pageserver B, then it is dangerous for A to do any object deletions, even of objects that it wrote
+itself, because pageserver's B metadata might reference those objects.
+
+To solve this without doing excessive synchronization, deletions are accumulated in deletion lists
+(also known as deadlists, or garbage lists), to be executed later. The actual execution of one of
+these lists has two preconditions:
+
+- That the executing node has written out the metadata index file since it decided to do
+  the deletion (i.e. no to-be-deleted files are referenced by the index)
+- That after writing that index file, the deleting node has confirmed that it is still the latest
+  generation (i.e. none of its attached tenants have been re-attached to another node).
+
+The order of operations for correctness is:
+
+1. Decide to delete one or more objects
+2. Write updated metadata (this means that if we are still the rightful leader, any subsequent leader will
```

Comment:
It were not question to me but let me answer:

> Do we prefix index part json with generational numbers?

Yes, there were examples in RFS: index_part.json-00000002-0001-00000001.
If index_part.json is not versioned then there will be race condition.

> Who's gonna take over?

As far as I understand, SK (safekeepers) will be somehow notified by control panel about current generation of PS and rejects (returns error or disconnect) any request from old PS. My question is what will happen with old PS which tenants do not perform any updates and so there is no message exchange between PS and SK.

Reply:
Thanks! That makes sense, +1, process of notifying safekeepers about pageserver changing its generation looks unclear to me.

Reply:
I've revised this to make the safekeeper change just an optional optimization, by adding a pageserver-side change to have it validate its generation with the control plane before updating remote_consistent_lsn.  There is a new "WAL trim changes" section above that describes the part necessary for correctness, and this section becomes "Optional Safekeeper changes"

In the updated text, I describe how we may include the generation number in pageserver->safekeeper messages, to propagate the generation numbers lazily (it can be lazy now that this is just an optimization)